### ISSUE TITLE
[discussion] DRA plugin raises UPDATE API request in scheduling cycle

### ISSUE THREAD

**Issue by Huang-Wei** [kubernetes/kubernetes#120502]
DRA (DynamicResourceAllocation) plugin tries to issue a synchronous API call in Reverse():

https://github.com/kubernetes/kubernetes/blob/c682d2b8c59f615d6896a1b1db22a5e7a5aaa23e/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L883

However, raising direct API requests in main scheduling cycle should be cautious as a blocking call may impact the overall scheduling perf & throughput. There are some solutions:

- avoid the UDPATE call totally
- change UPDATE to PATCH

Moreover, if we _have to_ use UPDATE, given UPDATE underneath compares versions and hence may fail on resource conflicts, we may pursue wrapping it in retry logic to avoid unnecessary churning in multiple scheduling cycles.

/sig scheduling
/kind feature

**Comment by Huang-Wei** [comment]
cc @pohly

**Comment by pohly** [comment]
> However, raising direct API requests in main scheduling cycle should be cautious as a blocking call may impact the overall scheduling perf & throughput.

Is the concern about impact on other pods which don't use claims or about throughput of claims that use claims?

It's understood that pods with claims will schedule more slowly. Whether it's fast enough is where we will need user feedback. I am actively looking for ways to make it faster (https://github.com/kubernetes/kubernetes/issues/113701).

The impact on other pods comes from blocking one worker goroutine, right? That's a valid concern. Perhaps the number of workers could get adjusted when some are dealing with "slow" pods, or simply increased overall?

> avoid the UDPATE call totally

To do that, the scheduling attempt would have to kick of the change in a goroutine, then abort the attempt, put the pod back into the queue, and continue when the change is complete. The event mechanism should support that, but it'll cause additional overhead.

In some cases, it also changes the behavior. For example, a ResourceClaim is only meant to be reserved while there really is a chance that the pod may get scheduled in the current attempt. If we kick of reserving the claims and later check, then it could happen that (for whatever reason) the pod that then holds the reservations doesn't get scheduled again (perhaps because some other plugin marked it as unschedulable) and then those ResourceClaims are "blocked" and cannot be used by some other pod either.

Error handling also becomes tricky. If we kick of some ResourceClaim change and put the pod into unschedulable waiting for a ResourceClaim change, then we also need to add an event for "that change failed".

What can be done without changing semantics or making the code more complex is to fan out multiple API calls and collect the results. This can be done whenever there is currently a for loop, like for the reserve case.

> change UPDATE to PATCH

For ResourceClaim, we have to check UID and ResourceVersion to ensure consistency. The same can be done with PATCH (one has to put those two fields into the patch and the apiserver will compare), but using PATCH then doesn't provide much benefit. I tried using SSA for PodSchedulingContext and found it to be slower because encoding is less efficient (JSON instead of protobuf) and because of the more complex server-side logic. See https://github.com/kubernetes/kubernetes/issues/113700.

> we may pursue wrapping it in retry logic to avoid unnecessary churning in multiple scheduling cycles.

If a ResourceClaim UPDATE fails because of a conflict, then starting a new full cycle is the right thing to do. All of the previous checks need to be repeated, otherwise it's unclear whether merely getting the latest ResourceClaim revision and applying the change again is still valid. A retry loop could try to detect "harmless" competing changes (say, someone added a label) and retry only in those cases, but that should be so rare that it doesn't warrant writing code for it.

I've also not seen conflicts occur much during stress testing. There was one occurrence which led me to https://github.com/kubernetes/kubernetes/pull/120334. For PodSchedulingContext, updating just some field regardless of changes in other fields is possible, so PATCH would have a conceptually advantage. Because UPDATE is still preferred for performance reasons, I implemented a fallback to PATCH exactly to avoid that extra pod scheduling cycle. But then when I tried to test it, I wasn't hitting the conflict at all anymore, which made me doubt whether it's worth the code, so I've not pursued that further.

**Comment by pohly** [comment]
> What can be done without changing semantics or making the code more complex is to fan out multiple API calls and collect the results. This can be done whenever there is currently a for loop, like for the reserve case.

I tried that. It did not really help the scheduler_perf test, quite the opposite, it got worse. But there isn't much IO latency in that test where concurrent API calls would help, so the extra work for the goroutines just causes overhead. It might be different in a real cluster. This needs more thought.

**Comment by Huang-Wei** [comment]
> Is the concern about impact on other pods which don't use claims or about throughput of claims that use claims?

It's more about the former case (pods that don't use claims). Pods that opt-in DRA are less concerned.

> The impact on other pods comes from blocking one worker goroutine, right?

Yes, if a Pod opt-in DRA spent excessive time on the sync UPDATE call, the Pods that are queued after (may not use DRA) are penalized as their earliest scheduling time are postponed.

> Perhaps the number of workers could get adjusted when some are dealing with "slow" pods, or simply increased overall?

We cannot actually. The scheduler's design is based on one worker, by which we sacrifice perf for accuracy.

> If a ResourceClaim UPDATE fails because of a conflict, then starting a new full cycle is the right thing to do.

My gut feeling (I may be wrong) is to put a question mark on this. As you mentioned, the resourceVersion change might be caused by some non-essentials fields' update.

**Comment by pohly** [comment]
> We cannot actually. The scheduler's design is based on one worker, by which we sacrifice perf for accuracy.

I see. Not sure how I missed that, but somehow I assumed that there would be multiple workers, as in other controllers. It definitely explains some lingering questions that I had about correctness of scheduling decisions, in particular regarding inter-pod relationships :hot_face: 

https://github.com/kubernetes/kubernetes/pull/120534 does the "avoid unnecessary churning in multiple scheduling cycles" for PodSchedulingContext. It should be useful regardless of future improvements.

It should be possible to move the PodSchedulingContext updates into a goroutine. I'm just not sure how to handle errors. If those calls fail, how can I get the pod from the unschedulable queue into the backoff queue? Is there perhaps already a special event for this? Should each update get its own goroutine or do you prefer a workqueue approach with a bounded number of goroutines?

ResourceClaim updates are harder. For `DeallocationRequested` I would just stick to the current inline approach. That should be rare. The common case is reserving for the pod and (less often) removing the reservation. One thought that I had was to let the scheduler check availability and then proceed based on the assumption that reservation will succeed, without actually updating the ResourceClaim object. Then the first thing that binding needs to do is to actually get all ResourceClaims reserved.

But isn't it so that there can be only one bind plugin? How can the dynamicresources plugin get its own code to run before whatever other plugin handles binding?

**Comment by pohly** [comment]
> If those calls fail, how can I get the pod from the unschedulable queue into the backoff queue? Is there perhaps already a special event for this?

This is open.

> Should each update get its own goroutine or do you prefer a workqueue approach with a bounded number of goroutines?

Binding uses one new goroutine per scheduling cycle, so I guess the same would be okay for `PodSchedulingContext` updates.

> But isn't it so that there can be only one bind plugin? How can the dynamicresources plugin get its own code to run before whatever other plugin handles binding?

`PreBind` is called for each plugin. This is where dynamicresources could to the actual reservation. If binding then fails, `handleBindingCycleError` calls `Unreserve`. This sounds doable.

**Comment by pohly** [comment]
> > If those calls fail, how can I get the pod from the unschedulable queue into the backoff queue? Is there perhaps already a special event for this?
>
> This is open.

@Huang-Wei pointed me towards https://github.com/kubernetes/kubernetes/blob/aa4ec3c5b006157751df4feac5c620c9dfadd2b7/pkg/scheduler/schedule_one.go#L317-L349 on Slack.

There are two issues with doing the same from a plugin:
- no access to that `internalqueue.AssignedPodDelete` event
- no access to the scheduling queue (not exposed via `framework.Handle`)

So framework API changes would be needed to make `PodSchedulingContext` updates robust when run in a goroutine. Doable, but not trivial. Same with reserve during binding.

Our overall conclusion was to focus on https://github.com/kubernetes/kubernetes/pull/120534 for 1.29 and revisit this topic once we know more about real clusters using DRA (workload mix, impact on scheduling, etc.).

@Huang-Wei: please correct me if I didn't capture this correctly.

**Comment by Huang-Wei** [comment]
> @Huang-Wei: please correct me if I didn't capture this correctly.

Yes, that's aligned with our offline discussion. #120534 looks promising to me. I will add more comments on that PR.

**Comment by alculquicondor** [comment]
> Our overall conclusion was to focus on https://github.com/kubernetes/kubernetes/pull/120534 for 1.29 and revisit this topic once we know more about real clusters using DRA (workload mix, impact on scheduling, etc.).

If we are targeting beta in 1.29, I hope we can have a design where we minimize the effect to pods not using DRA.

> ResourceClaim updates are harder. For DeallocationRequested I would just stick to the current inline approach. That should be rare.

This sounds reasonable. This is very akin to preemption, which also does API calls in PostFilter (API calls themselves are done in parallel).

> PreBind is called for each plugin. This is where dynamicresources could to the actual reservation. If binding then fails, handleBindingCycleError calls Unreserve. This sounds doable.

Bind plugins are also ordered, so you could consider that too. But perhaps PreBind is a better fit.

The important question is whether we can move the calls out of Reserve. For that, we need to evaluate whether that  could affect the correctness of scheduling decisions for the following pods.

**Comment by pohly** [comment]
Most of the blocking calls were moved in https://github.com/kubernetes/kubernetes/pull/121876.

@Huang-Wei: shall we consider this issue done?

**Comment by Huang-Wei** [comment]
> @Huang-Wei: shall we consider this issue done?

Yup, let's close it.
